src/main/java/org/cloudfoundry/promregator/rewrite/CFAllLabelsMetricFamilySamplesEnricher.java

```diff
@@ -24,44 +29,51 @@
 	public static final String LABELNAME_INSTANCE_NUMBER = "cf_instance_number";
 	
 	private static String[] labelNames = new String[] { LABELNAME_ORGNAME, LABELNAME_SPACENAME, LABELNAME_APPNAME, LABELNAME_INSTANCEID, LABELNAME_INSTANCE_NUMBER };
-	
 	public static String[] getEnrichingLabelNames() {
 		return labelNames.clone();
 	}
-	
+
 	private String orgName;
 	private String spaceName;
 	private String appName;
 	private String instanceId;
 
 	public CFAllLabelsMetricFamilySamplesEnricher(String orgName, String spaceName, String appName, String instanceId) {
-		this.instanceId = instanceId;
+		this.orgName = orgName;
 		this.spaceName = spaceName;
 		this.appName = appName;
-		this.orgName = orgName;
+		this.instanceId = instanceId;
 	}
 	
 	@Override
 	protected List<String> getEnrichedLabelNames(List<String> original) {
 		List<String> clone = new LinkedList<>(original);
+		for(String label: labelNames) {
+			clone.remove(label);
+			//TODO: log that label was removed
+		}
 		Collections.addAll(clone, labelNames);
-		
 		return clone;
 	}
-	
+
 	@Override
-	public List<String> getEnrichedLabelValues(List<String> original) {
-		List<String> clone = new LinkedList<>(original);
-		
-		clone.add(this.orgName);
-		clone.add(this.spaceName);
-		clone.add(this.appName);
-		clone.add(this.instanceId);
-		clone.add(getInstanceFromInstanceId(this.instanceId));
-		
-		return clone;
+	public List<String> getEnrichedLabelValues(List<String> originalLabelNames, List<String> originalLabelValues) {
+		List<String> labelListClone = new LinkedList<>(originalLabelValues);
+		Set<String> originalLabelNamesSet = new HashSet<>(originalLabelNames);
+		for(String label: labelNames) {
+			if(originalLabelNamesSet.contains(label)){
+				labelListClone.remove(originalLabelNames.indexOf(label));
```

Comment:
This is a double-lookup - and you will get an `O(n²)` algorithm again with it. 

BTW: Why are you removing it? It will be added anyhow afterwards. And the order of labels will be ignored by Prometheus. 

Perhaps an alternative idea, therefore, could be to "remember" (e.g. with five boolean variables) which labels you have already seen. Then, you could change the label's value directly in place - even without removing the value from the list before.

Just some thoughts...